doc/specs/#885 - Terminal Settings Model/Actions Addendum.md

```diff
@@ -192,43 +197,110 @@ structures. To get around this issue, we can introduce internal action IDs to en
- and save it to a new `_IDMap`.
+ `_KeyMap` is trivial. We can iterate over `_ActionList` to create a copy of every `Action` stored
+ and save it to a new `_ActionList`.
+
+### Layering Actions
+
+We need a way to determine where an action came from to minimize how many actions we serialize when we
+ write to disk. This is a two part approach that happens as we're loading the settings
+ 1. Load defaults.json
+     - For each of the actions in the JSON...
+         - Construct the `Command` (basically the `Command::LayerJson` we have today)
+         - Add it to the `ActionMap`
+           - this should update `_NameMap`, `_KeyMap`, and `_ActionList` appropriately
+           - if the newly added name/key chord conflicts with a pre-existing one,
+              redirect `_NameMap` or `_KeyMap` to the newly added `Command` instead,
+              and update the conflicting one.
+2. Load settings.json
+     - Add a parent to `ActionMap`
+         - The purpose of a parent is to continue a search when the current `ActionMap`
+            can't find a `Command` for a query. The parent is intended to be immutable.
+     - Load the actions array like normal (see step 1)
+
+Introducing a parent mechanism to `ActionMap` allows it to understand where a `Command`
+ came from. This allows us to minimize the number of actions we serialize when we write
+ to disk, as opposed to serializing the entire list of actions.
+
+`ActionMap` queries will need to check their parent when they cannot find a matching `Command`
+ in their `_ActionList` (or sooner).
+
+### Modifying Actions
+
+There are several ways a command can be modified:
+- change/remove the key chord
+- change the name
+- change the icon
+- change the action
+
+It is important that these modifications are done through `ActionMap` instead of `Command`.
+ This is to ensure that the `ActionMap` is always aligned with `Command`'s values. Thus,
+ we can add the following functions to `ActionMap`:
+
+ ```c++
+ runtimeclass ActionMap
+{
+    void SetKeyChord(Command cmd, KeyChord keys);
+    void SetName(Command cmd, String name);
+    void SetIcon(Command cmd, String iconPath);
+    void SetAction(Command cmd, ShortcutAction action, IActionArgs actionArgs);
+}
+ ```
 
-### Unbinding actions
+`SetKeyChord` will need to make sure to modify the `_KeyMap` and the provided `Command`.
+ If the new key chord was already taken, we also need to update the conflicting `Command`
+ and remove its key chord.
+
+`SetName` will need to make sure to modify the `_NameMap` and the provided `Command`.
+ We also need to iterate through the list of commands and see if the old name is still
+ in use by another command. If that's the case, don't remove `_NameMap[oldName]`,
+ instead retarget it.
 
-Removing a name or a key chord from an `Action` propagates the change to the `ActionMap`.
+`SetIcon` will only need to modify the provided `Command`.

```

Comment:
Does `SetIcon` really need to be exposed through the `ActionMap` then? Or can it just remain as a setter on the `Command` itself?

Reply:
It doesn't _need_ to be there, but it makes the API consistent.